keps/sig-instrumentation/1668-log-tracking/README.md

```diff
@@ -0,0 +1,372 @@
+  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
+  - [Monitoring Requirements](#monitoring-requirements)
+  - [Dependencies](#dependencies)
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+This KEP proposes a method for adding new three unique logging meta-data into K8s component logs.
+It makes us more easy to identify specific logs related to an user request (such as `kubectl apply`) and object (such as Pod, Deployment).
+It is expected to reduce investigation cost greatly when trouble shoothing.
+
+### New three unique logging meta-data
+
+We use three meta-data. These meta-data have different features and are used for troubleshooting from different perspectives.
+
+| meta-data name | feature |
+| ------ | ------ |
+| traceid | spans an user request. unique for user's request |
+| spanid | spans a controller action. unique for controller action |
+| initialtraceid | spans the entire object lifecycle. unique for related objects |
+
+### Note
+
+This KEP is **how** a component could add meta-data to logs. To actually add meta-data to K8s component logs, the following procedure is necessary in addition.
+- Open issues for each component, and discuss them with the SIGs that own that component.
+- After get agreement, utilize this KEP's feature to change the source code that outputs log to add meta-data into these logs.
+Please note that this KEP alone does not change the log format(does not add meta-data to logs).
+
+## Motivation
+
+Tracking logs among each Kubernetes component related to specific an user operation and objects is very tough work.
+It is necessary to match logs by basically using timestamps and object's name as hints.
+If multiple users throw many API requests at the same time, it is very difficult to track logs across each Kubernetes component log.
+
+### Goals
+
+ - Implement method which propagates new logging meta-data among each K8s component
```

Comment:
Is the goal to support arbitrary meta data? The KEP is mostly discussing "three unique logging meta-data" and webhooks (as example?), but there are several cases were additional context is necessary to understand a log message.

Currently code like https://github.com/kubernetes/kubernetes/blob/10744811f51cef741948676758563cfe04bb7ef2/pkg/controller/volume/persistentvolume/pv_controller.go#L413 makes sure that the object it operates on is repeated in each log message. Elsewhere I have seen ad-hoc log helper functions. But this breaks down when a function is called which then uses the global klog to print some output.

Structured logging with a logger that can accumulate additional key/value pairs (i.e. `WithValues` in https://pkg.go.dev/github.com/go-logr/logr#Logger) and passing that logger instance to all called functions would address that problem.

How to pass the logger instance was discussed in the original structured logging KEP, but then got removed when the scope got reduced to just structured output. I was hoping that this KEP here would address that gap.

FWIW, I think passing a logr instance via the context makes sense, with the global klog logger as fallback (via klogr). Logging in controller-runtime already follows that pattern (just with a different fallback) and once https://github.com/kubernetes-sigs/controller-runtime/issues/1275 is implemented, that code would be compatible also with other code that uses logr.